# 과제 2 - entropy 구하고, Gain 함수 구현하기 (lee hye min)

In [4]:
# 모듈 불러오기
import pandas as pd 
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/AugustLONG/ML01/master/01decisiontree/AllElectronics.csv')
df.drop("RID",axis=1, inplace = True) #RID는 그냥 Index라서 삭제

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 5 columns):
age                    14 non-null object
income                 14 non-null object
student                14 non-null object
credit_rating          14 non-null object
class_buys_computer    14 non-null object
dtypes: object(5)
memory usage: 640.0+ bytes


In [6]:
df

age  income student credit_rating class_buys_computer
0         youth    high      no          fair                  no
1         youth    high      no     excellent                  no
2   middle_aged    high      no          fair                 yes
3        senior  medium      no          fair                 yes
4        senior     low     yes          fair                 yes
5        senior     low     yes     excellent                  no
6   middle_aged     low     yes     excellent                 yes
7         youth  medium      no          fair                  no
8         youth     low     yes          fair                 yes
9        senior  medium     yes          fair                 yes
10        youth  medium     yes     excellent                 yes
11  middle_aged  medium      no     excellent                 yes
12  middle_aged    high     yes          fair                 yes
13       senior  medium      no     excellent                  no

In [7]:
new_feature = df["class_buys_computer"].value_counts()
yes = new_feature[0]
no = new_feature[1]
new_feature


    

yes    9
no     5
Name: class_buys_computer, dtype: int64

In [8]:
import math
def getEntropy(df, feature) :
    new_feature = df[feature].value_counts()
    yes = new_feature[0] #yes개수
    no = new_feature[1] #no개수
    count = len(df)
    entropy = ((-yes/count) *(math.log((yes/count),2)))+((-no/count) * (math.log((no/count),2)) )
    return(entropy)

In [9]:
getEntropy(df, "class_buys_computer")

0.9402859586706309

In [10]:
info_D = getEntropy(df, "class_buys_computer") # 목표변수에 대한 Entropy 를 구하기
columns = list(df.loc[:, df.columns != "class_buys_computer"]) 
#['age', 'income', 'student', 'credit_rating']
    
age = df[columns[0]].value_counts() #나이
income = df[columns[1]].value_counts() #수입
student = df[columns[2]].value_counts() #신분
credit_rating = df[columns[3]].value_counts() #연봉

In [11]:
def getGainA(df, feature , target ) : 
    
    entropy=0
    for i in pd.unique(df[feature]).tolist():
        # 원하는 속성을 골라냄
        df2 = df[df[feature].apply(lambda x: x in i)]
    
        # target 변수가 참인 개수
        label_p = len(df2[df2[target]=='yes'])
        
        # 전체 class 중에 선택된 클래스의 확률
        prob = len(df2)/len(df)
        
        # 뽑힌 class 에서 target 변수가 참인 확률 
        p = label_p/len(df2)
        
        try : 
            entropy -= prob* ((-p)*(math.log(p,2)) -(1-p)*(math.log((1-p),2)))
        except : 
            if p== 0:    # p==0 이면 계산이 log 0 이라 예외처리 해줌
                entropy -= prob * (-(1-p)*(math.log((1-p),2)))
            elif p==1 :  # p==1 이면 계산이 log 0 이라 예외처리 해줌
                entropy -= prob *( -p*(math.log(p,2)))
       
    print(-entropy)
    return(-entropy)


In [12]:
getGainA(df, "age","class_buys_computer")

0.6935361388961918


0.6935361388961918

In [14]:
    info_D = getEntropy(df,"class_buys_computer") 
    info_D_age = getGainA(df,'age', 'class_buys_computer')
    info_D_income = getGainA(df,'income', 'class_buys_computer')
    info_D_student = getGainA(df,'student', 'class_buys_computer')
    info_D_credit = getGainA(df,'credit_rating', 'class_buys_computer')

0.6935361388961918
0.9110633930116763
0.7884504573082896
0.8921589282623617


In [16]:
gain_age = info_D - info_D_age
gain_income = info_D - info_D_income
gain_student = info_D - info_D_student
gain_credit = info_D - info_D_credit

In [17]:
gain_age, gain_income, gain_student, gain_credit

(0.2467498197744391,
 0.029222565658954647,
 0.15183550136234136,
 0.04812703040826927)

정보 획득이 가장 높은 변수는 age이며, 획득량은 0.2467498197744391 이다.